clang16 warnings

Message from Brian:

> clang 16 is being prepared for release in early March, so it is time to
> start reporting its issues.  See
> https://www.stats.ox.ac.uk/pub/bdr/clang16/
>
>
> Shortly these logs will appear as 'clang16' additional issues.
> 
> Please correct before 2023-02-16 to safely retain your package on CRAN.
> (Some have much closer deadlines for other issues.)

README.txt
> Tests as for fedora-clang but using clang 16.0.0git rather than 15.0.7.
> This is scheduled for release on Mar 7th.
> 
> Currently using C23 as the default C standard (but C23 issues are
> reported elsewhare).
> 
> Other details as
> https://www.stats.ox.ac.uk/pub/bdr/Rconfig/r-devel-linux-x86_64-fedora-clang
> 
> References for Rf_length in system headers come from includeing R headers
> before system headers rather than after: see 'Writing R Extnesions'.
> 
> The [-Wenum-constexpr-conversion] errors seen in
> 
> RSQLite
> 
> are from the old Boost headers.
> 
> std:unary_function was deprecated in C+11 and removed in C++17, and finally
> in clang 16 in C++17 mode.  Boost 1.80 was still using it.
> 


RMixtCompIO.out
> * using log directory ‘/data/gannet/ripley/R/packages/tests-clang-trunk/RMixtCompIO.Rcheck’
> * using R Under development (unstable) (2023-01-20 r83646)
> * using platform: x86_64-pc-linux-gnu (64-bit)
> * R was compiled by
>     clang version 16.0.0 (https://github.com/llvm/llvm-project.git 2133e8b9f942f91ec54e28c580fccf6d6b26c62e)
>     GNU Fortran (GCC) 12.2.1 20221121 (Red Hat 12.2.1-4)
> * running under: Fedora Linux 36 (Workstation Edition)
> * using session charset: UTF-8
> * using option ‘--no-stop-on-test-error’
> * checking for file ‘RMixtCompIO/DESCRIPTION’ ... OK
> * checking extension type ... Package
> * this is package ‘RMixtCompIO’ version ‘4.0.8’
> * package encoding: UTF-8
> * checking package namespace information ... OK
> * checking package dependencies ... OK
> * checking if this is a source package ... OK
> * checking if there is a namespace ... OK
> * checking for executable files ... OK
> * checking for hidden files and directories ... OK
> * checking for portable file names ... OK
> * checking for sufficient/correct file permissions ... OK
> * checking whether package ‘RMixtCompIO’ can be installed ... [454s/374s] ERROR
> Installation failed.
> See ‘/data/gannet/ripley/R/packages/tests-clang-trunk/RMixtCompIO.Rcheck/00install.out’ for details.
> * DONE
> 
> Status: 1 ERROR
> See
>   ‘/data/gannet/ripley/R/packages/tests-clang-trunk/RMixtCompIO.Rcheck/00check.log’
> for details.
> 
> Command exited with non-zero status 1
> Time 6:16.95, 436.39 + 19.82

[RMixtCompIO.log](https://github.com/modal-inria/MixtComp/files/10466557/RMixtCompIO.log)

the error seems to come from the sort overload for custom matrix
https://github.com/modal-inria/MixtComp/blob/master/MixtComp/src/lib/LinAlg/EigenMatrixBaseAddons.h#L237-L240

> In file included from lib/Mixture/Functional/FunctionCS.cpp:23:
> In file included from lib/Mixture/Functional/FuncCSClass.h:26:
> In file included from lib/LinAlg/LinAlg.h:26:
> In file included from lib/LinAlg/Typedef.h:27:
> In file included from /usr/local/clang-trunk/bin/../include/c++/v1/string:576:
> In file included from /usr/local/clang-trunk/bin/../include/c++/v1/string_view:1025:
> In file included from /usr/local/clang-trunk/bin/../include/c++/v1/algorithm:1771:
> In file included from /usr/local/clang-trunk/bin/../include/c++/v1/__algorithm/nth_element.h:15:
> /usr/local/clang-trunk/bin/../include/c++/v1/__algorithm/sort.h:656:11: error: no viable overloaded '-='
>     __lm1 -= (__right_bitset == 0) ? difference_type(__detail::__block_size) : difference_type(0);
>     ~~~~~ ^  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> /usr/local/clang-trunk/bin/../include/c++/v1/__algorithm/sort.h:859:20: note: in instantiation of function template specialization 'std::__bitset_partition<std::_ClassicAlgPolicy, Eigen::MatrixBase<Eigen::Matrix<double, -1, 1, 0>>::Iterator, std::__less<double> &>' requested here
>             ? std::__bitset_partition<_AlgPolicy, _RandomAccessIterator, _Compare>(__first, __last, __comp)
>                    ^
> /usr/local/clang-trunk/bin/../include/c++/v1/__algorithm/sort.h:919:8: note: in instantiation of function template specialization 'std::__introsort<std::_ClassicAlgPolicy, std::__less<double> &, Eigen::MatrixBase<Eigen::Matrix<double, -1, 1, 0>>::Iterator, false>' requested here
>   std::__introsort<_AlgPolicy,
>        ^
> /usr/local/clang-trunk/bin/../include/c++/v1/__algorithm/sort.h:983:10: note: in instantiation of function template specialization 'std::__sort<std::__less<double> &, Eigen::MatrixBase<Eigen::Matrix<double, -1, 1, 0>>::Iterator>' requested here
>     std::__sort<_WrappedComp>(std::__unwrap_iter(__first), std::__unwrap_iter(__last), __wrapped_comp);
>          ^
> /usr/local/clang-trunk/bin/../include/c++/v1/__algorithm/sort.h:990:8: note: in instantiation of function template specialization 'std::__sort_impl<std::_ClassicAlgPolicy, Eigen::MatrixBase<Eigen::Matrix<double, -1, 1, 0>>::Iterator, std::__less<double>>' requested here
>   std::__sort_impl<_ClassicAlgPolicy>(std::move(__first), std::move(__last), __comp);
>        ^
> /usr/local/clang-trunk/bin/../include/c++/v1/__algorithm/sort.h:996:8: note: in instantiation of function template specialization 'std::sort<Eigen::MatrixBase<Eigen::Matrix<double, -1, 1, 0>>::Iterator, std::__less<double>>' requested here
>   std::sort(__first, __last, __less<typename iterator_traits<_RandomAccessIterator>::value_type>());
>        ^
> lib/LinAlg/EigenMatrixBaseAddons.h:239:7: note: in instantiation of function template specialization 'std::sort<Eigen::MatrixBase<Eigen::Matrix<double, -1, 1, 0>>::Iterator>' requested here
>         std::sort(derived().begin(), derived().end());
>              ^
> lib/Mixture/Functional/FunctionCS.cpp:237:10: note: in instantiation of member function 'Eigen::MatrixBase<Eigen::Matrix<double, -1, 1, 0>>::sort' requested here
>         sortedT.sort();

--- a/MixtComp/.clang-format
+++ b/MixtComp/.clang-format
@@ -1,0 +1,14 @@
+# https://clang.llvm.org/docs/ClangFormatStyleOptions.html
+# https://llvm.org/docs/CodingStandards.html
+
+BasedOnStyle: LLVM
+
+AccessModifierOffset: -4
+AllowShortIfStatementsOnASingleLine: false
+ColumnLimit: 0
+FixNamespaceComments: false
+IndentCaseLabels: false
+IndentWidth: 4
+NamespaceIndentation: All
+TabWidth: 4
+UseTab: Always

--- a/MixtComp/README.md
+++ b/MixtComp/README.md
@@ -29,7 +29,7 @@ brew install cmake boost eigen lcov
 
 ## Windows
 
-To install the necessary tools, user needs to install [cygwin](https://www.cygwin.com/). gcc, cmake, boost and eigen must be installed through the cygwin packages installer.
+To install the necessary tools, user needs to install [cygwin](https://www.cygwin.com/). gcc, cmake, boost and eigen must be installed through the cygwin packages installer. Note: WSL might be a better solution, but does it allows to compare with CRAN builds for example ?
 
 ## Compilation & tests
 

--- a/MixtComp/TODO.md
+++ b/MixtComp/TODO.md
@@ -31,7 +31,7 @@
 ## Performances
 
 - remove regex in data parsing. Will be faster but less tolerant to data format.
-- Statistics object generate a boost::variate_generator each time a variable is sampled. It should be possible to generate a vector of value, when the parameters do not vary.
+- Statistics object generate a boost::variate_generator each time a variable is sampled. It should be possible to generate a vector of value, when the parameters do not vary. -> Use standard library for sampling, as in Multinomial Statistics.
 
 ## Long Term
 

--- a/MixtComp/docs/Graph.md
+++ b/MixtComp/docs/Graph.md
@@ -6,93 +6,88 @@ Inputs and outputs are given as Graph objects. MixtComp methods are templated wi
 - R objects (*RGraph* class, cf. [RMixtCompIO](../../RMixtCompIO/src/RGraph.h))
 - python objects (*PyGraph* class, cf. [pyMixtComp](../../pyMixtComp/src/lib/PyGraph.h))
 
-
 The different *Graph* classes must implement the following method:
 
 - `void getSubGraph(const std::vector<std::string>& path, Graph& j) const;`: get the Graph at the given path. It modifies the `j` argument.
 - `template<typename Type> void add_payload(const std::vector<std::string>& path, const std::string& name, const Type& p);`: add an object in the Graph and create the complete path to it if it does not exist yet.
-- `void addSubGraph(const std::vector<std::string>& path, const std::string& name, const Graph& p);`: add an other Graph in the Graph and create the complete path to it if it does not exist yet. 
+- `void addSubGraph(const std::vector<std::string>& path, const std::string& name, const Graph& p);`: add an other Graph in the Graph and create the complete path to it if it does not exist yet.
 - `template<typename Type> Type get_payload(const std::vector<std::string>& path, const std::string& name) const;`: get the element named `name` in `path`
 - `template<typename Type> void get_payload(const std::vector<std::string>& path, const std::string& name, Type& p);`: get the element named `name` in `path`. This version modifies the `p` argument.
 - `bool exist_payload(const std::vector<std::string>& path, const std::string& name) const;`: test if an element named `name` exists in `path`.
 - `void name_payload(const std::vector<std::string>& path, std::list<std::string>& l) const;`: get the names of all the elements in `path` It modifies the `l` argument.
 
-with `path` a vector containing the path of the element to reach or save an element (e.g. ["variable", "param", "var1", "stat"] the element in the following path: variable/param/var1/stat). `Type` can be `int`, `double`, `bool`, `vector<T>`, `vector<vector<T> >`, `NamedVector<T>`, `NamedMatrix<T>`. `NamedMatrix<T>` is a struct containing an Eigen Matrix `mat_` and two vector<string> named `colnames_` and `rownames_`. `NamedVector<T>` is a struct containing an Eigen Matrix with 1 column `vec_` and a vector<string> named `rownames_`  (see [NamedAlgebra](../src/lib/IO/NamedAlgebra.h) and [LinAlg](../src/lib/LinAlg/LinAlg.h) for mode details about `NamedVector` and `NamedMatrix`). 
+with `path` a vector containing the path of the element to reach or save an element (e.g. ["variable", "param", "var1", "stat"] the element in the following path: variable/param/var1/stat). `Type` can be `int`, `double`, `bool`, `vector<T>`, `vector<vector<T> >`, `NamedVector<T>`, `NamedMatrix<T>`. `NamedMatrix<T>` is a struct containing an Eigen Matrix `mat_` and two `vector<string>` named `colnames_` and `rownames_`. `NamedVector<T>` is a struct containing an Eigen Matrix with 1 column `vec_` and a `vector<string>` named `rownames_`  (see [NamedAlgebra](../src/lib/IO/NamedAlgebra.h) and [LinAlg](../src/lib/LinAlg/LinAlg.h) for mode details about `NamedVector` and `NamedMatrix`).
 
-
-
-# Structure of input Graphs
+## Structure of input Graphs
 
 Three Graphs for learning (and four for prediction)  (see [file](./dataFormat.md) for mdoe details):
 
 - algo
-    - nClass
-    - nInd
-    - nbBurnInIter
-    - nbIter
-    - nbGibbsBurnInIter
-    - nbGibbsIter
-    - nInitPerClass
-    - nSemTry
-    - ratioStableCriterion
-    - nStableCriterion
-    - confidenceLevel
-    - mode ("learn" or "predict")
+  - nClass
+  - nInd
+  - nbBurnInIter
+  - nbIter
+  - nbGibbsBurnInIter
+  - nbGibbsIter
+  - nInitPerClass
+  - nSemTry
+  - ratioStableCriterion
+  - nStableCriterion
+  - confidenceLevel
+  - mode ("learn" or "predict")
 
 - data
-    - var1 (array of string)
-    - var2
-    - ...
+  - var1 (array of string)
+  - var2
+  - ...
 
 - desc
-    - var1
-        - type
-        - paramStr
-    - var2
-    - ...
+  - var1
+    - type
+    - paramStr
+  - var2
+  - ...
 
 - param (output Graph of a MixtComp run. Only for prediction, see below)
 
-
-# Structure of output Graph
+## Structure of output Graph
 
 An unique Graph containing: (see [file](./objectOutput.md) for more details)
 
 - algo (copy of algo input)
-- mixture 
+- mixture
 - variable
-    - type
-        - var1: type of var1
-        - var2: type of var2
-    - data
-        - var1
-            - completed
-            - stat (to be implemented, except for LatentClass -> immediate implementation)
-        - var2
-        - ...
-    - param
-        - var1 (case 1: simple model)
-            - log 
-            - stat
-            - paramStr
-        - var2 (case 2: functional)
-            - alpha
-                - log 
-                - stat               
-            - beta
-                - log 
-                - stat 
-            - sd
-                - log 
-                - stat 
-            - paramStr
-        - var3 (case 2: rank)
-            - mu
-                - log 
-                - stat               
-            - pi
-                - log 
-                - stat 
-            - paramStr 
-        - ...
-
+  - type
+    - var1: type of var1
+    - var2: type of var2
+  - data
+    - var1
+      - completed
+      - stat (to be implemented, except for LatentClass -> immediate implementation)
+    - var2
+    - ...
+  - param
+    - var1 (case 1: simple model)
+      - log
+      - stat
+      - paramStr
+    - var2 (case 2: functional)
+      - alpha
+        - log
+        - stat
+      - beta
+        - log
+        - stat
+      - sd
+        - log
+        - stat
+      - paramStr
+    - var3 (case 2: rank)
+      - mu
+        - log
+        - stat
+      - pi
+        - log
+        - stat
+      - paramStr
+    - ...

--- a/MixtComp/docs/developmentEnvironment.md
+++ b/MixtComp/docs/developmentEnvironment.md
@@ -1,0 +1,15 @@
+# Development environment
+
+## Current
+
+VS Code with the the following extensions:
+
+- clangd
+  - Disable Intellisense from Microsoft C++ extension. The formatting should be automatically handled upon detection of the `.clang-format` file at the root of the project.
+- markdownlint
+
+Enable the "Format on save" option to ensure everything is properly formatted.
+
+## Previous
+
+The general style used is K&R from Eclipse, modified for a line wrap at 200 characters.

--- a/MixtComp/docs/styleguide.md
+++ b/MixtComp/docs/styleguide.md
@@ -1,5 +1,0 @@
-# Styleguide
-
-This document describes the conventions used to code. They might not have been enforced this the beginning of the project, but should be for any new code.
-
-The general style used is K&R from Eclipse, modified for a line wrap at 200 characters.

--- a/MixtComp/src/lib/LinAlg/ConstIterator.h
+++ b/MixtComp/src/lib/LinAlg/ConstIterator.h
@@ -140,6 +140,18 @@ public:
 		return *this;
 	}
 
+	const ConstIterator& operator++(int) {
+		if (i_ < rows_ - 1) // row increment
+				{
+			++i_;
+		} else // column increment
+		{
+			i_ = 0;
+			++j_;
+		}
+		return *this;
+	}
+
 	const ConstIterator& operator--() {
 		if (i_ > 0) {
 			--i_;

--- a/MixtComp/src/lib/LinAlg/ConstIterator.h
+++ b/MixtComp/src/lib/LinAlg/ConstIterator.h
@@ -47,7 +47,7 @@ public:
 		: i_(i), j_(j), rows_(mat.rows()), p_mat_(&mat) {
 	}
 
-	ConstIterator(Iterator &other)
+	ConstIterator(const Iterator &other)
 		: i_(other.i_),
 		  j_(other.j_),
 		  rows_(other.rows_),

--- a/MixtComp/src/lib/LinAlg/ConstIterator.h
+++ b/MixtComp/src/lib/LinAlg/ConstIterator.h
@@ -28,8 +28,6 @@
 //#include <iostream>
 //#include <iterator>
 
-// #include <iterator>
-
 class ConstIterator {
 
 public:
@@ -39,6 +37,11 @@ public:
 	using difference_type = int;
 	using pointer = Scalar *;
 	using reference = Scalar &;
+
+	int i_;
+	int j_;
+	int rows_;
+	const Derived *p_mat_;
 
 	ConstIterator(int i, int j, const Derived &mat)
 		: i_(i), j_(j), rows_(mat.rows()), p_mat_(&mat) {
@@ -174,11 +177,6 @@ public:
 		}
 		return *this;
 	}
-
-	int i_;
-	int j_;
-	int rows_;
-	const Derived *p_mat_;
 
 	void posToIn(int pos, int &i, int &j) const {
 		std::div_t divresult;

--- a/MixtComp/src/lib/LinAlg/ConstIterator.h
+++ b/MixtComp/src/lib/LinAlg/ConstIterator.h
@@ -28,6 +28,8 @@
 //#include <iostream>
 //#include <iterator>
 
+#include <iterator>
+
 class ConstIterator {
 public:
 	// iterator traits

--- a/MixtComp/src/lib/LinAlg/ConstIterator.h
+++ b/MixtComp/src/lib/LinAlg/ConstIterator.h
@@ -47,6 +47,12 @@ public:
 		: i_(i), j_(j), rows_(mat.rows()), p_mat_(&mat) {
 	}
 
+	ConstIterator(Iterator &other)
+		: i_(other.i_),
+		  j_(other.j_),
+		  rows_(other.rows_),
+		  p_mat_(other.p_mat_) {}
+
 	ConstIterator operator+(int i) {
 		int posP, iP, jP;
 		posP = pos();

--- a/MixtComp/src/lib/LinAlg/ConstIterator.h
+++ b/MixtComp/src/lib/LinAlg/ConstIterator.h
@@ -47,13 +47,6 @@ public:
 		: i_(i), j_(j), rows_(mat.rows()), p_mat_(&mat) {
 	}
 
-	ConstIterator(const Iterator &it)
-		: i_(it.i_),
-		  j_(it.j_),
-		  rows_(it.rows_),
-		  p_mat_(it.p_mat_) {
-	}
-
 	ConstIterator operator+(int i) {
 		int posP, iP, jP;
 		posP = pos();

--- a/MixtComp/src/lib/LinAlg/ConstIterator.h
+++ b/MixtComp/src/lib/LinAlg/ConstIterator.h
@@ -31,6 +31,7 @@
 #include <iterator>
 
 class ConstIterator {
+
 public:
 	// iterator traits
 	using iterator_category = std::random_access_iterator_tag;

--- a/MixtComp/src/lib/LinAlg/ConstIterator.h
+++ b/MixtComp/src/lib/LinAlg/ConstIterator.h
@@ -28,7 +28,7 @@
 //#include <iostream>
 //#include <iterator>
 
-#include <iterator>
+// #include <iterator>
 
 class ConstIterator {
 

--- a/MixtComp/src/lib/LinAlg/ConstIterator.h
+++ b/MixtComp/src/lib/LinAlg/ConstIterator.h
@@ -28,15 +28,18 @@
 //#include <iostream>
 //#include <iterator>
 
-class ConstIterator: public std::iterator<std::random_access_iterator_tag,
+class ConstIterator : public std::iterator<std::random_access_iterator_tag,
 		Scalar, int, Scalar*, Scalar&> {
 public:
-	ConstIterator(int i, int j, const Derived& mat) :
-			i_(i), j_(j), rows_(mat.rows()), p_mat_(&mat) {
+	ConstIterator(int i, int j, const Derived& mat)
+	: i_(i), j_(j), rows_(mat.rows()), p_mat_(&mat) {
 	}
 
-	ConstIterator(const Iterator& it) :
-			i_(it.i_), j_(it.j_), rows_(it.rows_), p_mat_(it.p_mat_) {
+	ConstIterator(const Iterator& it)
+	: i_(it.i_),
+	  j_(it.j_),
+	  rows_(it.rows_),
+	  p_mat_(it.p_mat_) {
 	}
 
 	ConstIterator operator+(int i) {
@@ -47,12 +50,12 @@ public:
 		return ConstIterator(iP, jP, *p_mat_);
 	}
 
-	ConstIterator& operator+=(int i) {
+	ConstIterator operator+=(int i) {
 		posToIn(pos() + i, i_, j_);
 		return *this;
 	}
 
-	ConstIterator& operator-=(int i) {
+	ConstIterator operator-=(int i) {
 		posToIn(pos() - i, i_, j_);
 		return *this;
 	}
@@ -140,7 +143,8 @@ public:
 		return *this;
 	}
 
-	const ConstIterator& operator++(int) {
+	const ConstIterator operator++(int) {
+		ConstIterator temp(*this);
 		if (i_ < rows_ - 1) // row increment
 				{
 			++i_;
@@ -149,7 +153,7 @@ public:
 			i_ = 0;
 			++j_;
 		}
-		return *this;
+		return temp;
 	}
 
 	const ConstIterator& operator--() {

--- a/MixtComp/src/lib/LinAlg/ConstIterator.h
+++ b/MixtComp/src/lib/LinAlg/ConstIterator.h
@@ -52,6 +52,11 @@ public:
 		return *this;
 	}
 
+	ConstIterator& operator-=(int i) {
+		posToIn(pos() - i, i_, j_);
+		return *this;
+	}
+
 	ConstIterator operator-(int i) {
 		int posP, iP, jP;
 		posP = pos();

--- a/MixtComp/src/lib/LinAlg/ConstIterator.h
+++ b/MixtComp/src/lib/LinAlg/ConstIterator.h
@@ -34,18 +34,18 @@ public:
 	using iterator_category = std::random_access_iterator_tag;
 	using value_type = Scalar;
 	using difference_type = int;
-	using pointer = Scalar*;
-	using reference = Scalar&;
+	using pointer = Scalar *;
+	using reference = Scalar &;
 
-	ConstIterator(int i, int j, const Derived& mat)
-	: i_(i), j_(j), rows_(mat.rows()), p_mat_(&mat) {
+	ConstIterator(int i, int j, const Derived &mat)
+		: i_(i), j_(j), rows_(mat.rows()), p_mat_(&mat) {
 	}
 
-	ConstIterator(const Iterator& it)
-	: i_(it.i_),
-	  j_(it.j_),
-	  rows_(it.rows_),
-	  p_mat_(it.p_mat_) {
+	ConstIterator(const Iterator &it)
+		: i_(it.i_),
+		  j_(it.j_),
+		  rows_(it.rows_),
+		  p_mat_(it.p_mat_) {
 	}
 
 	ConstIterator operator+(int i) {
@@ -74,44 +74,44 @@ public:
 		return ConstIterator(iP, jP, *p_mat_);
 	}
 
-	int operator-(const ConstIterator& it) {
+	int operator-(const ConstIterator &it) {
 		return pos() - it.pos();
 	}
 
-	bool operator<(const ConstIterator& it) {
+	bool operator<(const ConstIterator &it) {
 		if (j_ < it.j_) {
 			return true;
 		} else if (j_ > it.j_) {
 			return false;
 		} else if (i_ < it.i_) // in this case, j_ == it.j_, hence comparison must be done on i_
-				{
+		{
 			return true;
 		}
 
 		return false;
 	}
 
-	bool operator==(const ConstIterator& it) {
+	bool operator==(const ConstIterator &it) {
 		if (i_ == it.i_ && j_ == it.j_)
 			return true;
 		else
 			return false;
 	}
 
-	bool operator>(const ConstIterator& it) {
+	bool operator>(const ConstIterator &it) {
 		if (j_ > it.j_) {
 			return true;
 		} else if (j_ > it.j_) {
 			return false;
 		} else if (i_ > it.i_) // in this case, j_ == it.j_, hence comparison must be done on i_
-				{
+		{
 			return true;
 		}
 
 		return false;
 	}
 
-	bool operator>=(const ConstIterator& it) {
+	bool operator>=(const ConstIterator &it) {
 		if (!(*this < it)) {
 			return true;
 		}
@@ -119,7 +119,7 @@ public:
 		return false;
 	}
 
-	bool operator!=(const ConstIterator& it) {
+	bool operator!=(const ConstIterator &it) {
 		if (i_ != it.i_ || j_ != it.j_)
 			return true;
 		else
@@ -129,17 +129,17 @@ public:
 	/**
 	 * returns const Scalar instead of const Scalar& to avoid error with temporaries
 	 */
-	const Scalar& operator*() const {
+	const Scalar &operator*() const {
 		return (*p_mat_)(i_, j_);
 	}
 
-	const Scalar* operator->() const {
+	const Scalar *operator->() const {
 		return &(*p_mat_)(i_, j_);
 	}
 
-	const ConstIterator& operator++() {
+	const ConstIterator &operator++() {
 		if (i_ < rows_ - 1) // row increment
-				{
+		{
 			++i_;
 		} else // column increment
 		{
@@ -152,7 +152,7 @@ public:
 	const ConstIterator operator++(int) {
 		ConstIterator temp(*this);
 		if (i_ < rows_ - 1) // row increment
-				{
+		{
 			++i_;
 		} else // column increment
 		{
@@ -162,7 +162,7 @@ public:
 		return temp;
 	}
 
-	const ConstIterator& operator--() {
+	const ConstIterator &operator--() {
 		if (i_ > 0) {
 			--i_;
 		} else {
@@ -175,9 +175,9 @@ public:
 	int i_;
 	int j_;
 	int rows_;
-	const Derived* p_mat_;
+	const Derived *p_mat_;
 
-	void posToIn(int pos, int& i, int& j) const {
+	void posToIn(int pos, int &i, int &j) const {
 		std::div_t divresult;
 		divresult = std::div(pos, rows_);
 

--- a/MixtComp/src/lib/LinAlg/ConstIterator.h
+++ b/MixtComp/src/lib/LinAlg/ConstIterator.h
@@ -82,8 +82,7 @@ public:
 			return true;
 		} else if (j_ > it.j_) {
 			return false;
-		} else if (i_ < it.i_) // in this case, j_ == it.j_, hence comparison must be done on i_
-		{
+		} else if (i_ < it.i_) { // in this case, j_ == it.j_, hence comparison must be done on i_
 			return true;
 		}
 
@@ -102,8 +101,7 @@ public:
 			return true;
 		} else if (j_ > it.j_) {
 			return false;
-		} else if (i_ > it.i_) // in this case, j_ == it.j_, hence comparison must be done on i_
-		{
+		} else if (i_ > it.i_) { // in this case, j_ == it.j_, hence comparison must be done on i_
 			return true;
 		}
 

--- a/MixtComp/src/lib/LinAlg/ConstIterator.h
+++ b/MixtComp/src/lib/LinAlg/ConstIterator.h
@@ -28,9 +28,15 @@
 //#include <iostream>
 //#include <iterator>
 
-class ConstIterator : public std::iterator<std::random_access_iterator_tag,
-		Scalar, int, Scalar*, Scalar&> {
+class ConstIterator {
 public:
+	// iterator traits
+	using iterator_category = std::random_access_iterator_tag;
+	using value_type = Scalar;
+	using difference_type = int;
+	using pointer = Scalar*;
+	using reference = Scalar&;
+
 	ConstIterator(int i, int j, const Derived& mat)
 	: i_(i), j_(j), rows_(mat.rows()), p_mat_(&mat) {
 	}

--- a/MixtComp/src/lib/LinAlg/Iterator.h
+++ b/MixtComp/src/lib/LinAlg/Iterator.h
@@ -26,8 +26,8 @@
 // reference on what needs to be implemented for STL iterators:
 // http://stackoverflow.com/questions/7758580/writing-your-own-stl-container/7759622#7759622
 
-// #include <iostream>
-// #include <iterator>
+//#include <iostream>
+//#include <iterator>
 
 /**
  * Iterator class for Eigen::Matrix. This file is included from
@@ -36,154 +36,146 @@
  * mixt::Matrix.
  */
 
-class Iterator {
+class Iterator : public std::iterator<
+					 std::random_access_iterator_tag, Scalar,
+					 int, Scalar *, Scalar &> {
 public:
-  // iterator traits
-  using iterator_category = std::random_access_iterator_tag;
-  using value_type = Scalar;
-  using difference_type = int;
-  using pointer = Scalar*;
-  using reference = Scalar&;
+	Iterator(int i, int j, Derived &mat)
+		: i_(i),
+		  j_(j),
+		  rows_(mat.rows()),
+		  p_mat_(&mat) {}
 
-  Iterator(int i, int j, Derived &mat)
-      : i_(i), j_(j), rows_(mat.rows()), p_mat_(&mat) {}
+	Iterator operator+(int i) {
+		int posP, iP, jP;
+		posP = pos();
+		posP += i;
+		posToIn(posP, iP, jP);
+		return Iterator(iP, jP, *p_mat_);
+	}
 
-  Iterator(const Iterator &other)
-      : i_(other.i_), j_(other.j_), rows_(other.rows_), p_mat_(other.p_mat_) {}
+	Iterator &operator+=(int i) {
+		posToIn(pos() + i, i_, j_);
+		return *this;
+	}
 
-  Iterator operator+(int i) {
-    int posP, iP, jP;
-    posP = pos();
-    posP += i;
-    posToIn(posP, iP, jP);
-    return Iterator(iP, jP, *p_mat_);
-  }
+	Iterator &operator-=(int i) {
+		posToIn(pos() - i, i_, j_);
+		return *this;
+	}
 
-  Iterator &operator+=(int i) {
-    posToIn(pos() + i, i_, j_);
-    return *this;
-  }
+	Iterator operator-(int i) {
+		int posP, iP, jP;
+		posP = pos();
+		posP -= i;
+		posToIn(posP, iP, jP);
+		return Iterator(iP, jP, *p_mat_);
+	}
 
-  Iterator &operator-=(int i) {
-    posToIn(pos() - i, i_, j_);
-    return *this;
-  }
+	int operator-(const Iterator &it) { return pos() - it.pos(); }
 
-  Iterator operator-(int i) {
-    int posP, iP, jP;
-    posP = pos();
-    posP -= i;
-    posToIn(posP, iP, jP);
-    return Iterator(iP, jP, *p_mat_);
-  }
+	bool operator<(const Iterator &it) {
+		if (j_ < it.j_) {
+			return true;
+		} else if (j_ > it.j_) {
+			return false;
+		} else if (i_ < it.i_) // in this case, j_ == it.j_, hence comparison must
+							   // be done on i_
+		{
+			return true;
+		}
 
-  int operator-(const Iterator &it) { return pos() - it.pos(); }
+		return false;
+	}
 
-  bool operator<(const Iterator &it) {
-    if (j_ < it.j_) {
-      return true;
-    } else if (j_ > it.j_) {
-      return false;
-    } else if (i_ < it.i_) // in this case, j_ == it.j_, hence comparison must
-                           // be done on i_
-    {
-      return true;
-    }
+	bool operator==(const Iterator &it) {
+		if (i_ == it.i_ && j_ == it.j_)
+			return true;
+		else
+			return false;
+	}
 
-    return false;
-  }
+	bool operator>(const Iterator &it) {
+		if (j_ > it.j_) {
+			return true;
+		} else if (j_ > it.j_) {
+			return false;
+		} else if (i_ > it.i_) // in this case, j_ == it.j_, hence comparison must
+							   // be done on i_
+		{
+			return true;
+		}
 
-  bool operator==(const Iterator &it) {
-    if (i_ == it.i_ && j_ == it.j_)
-      return true;
-    else
-      return false;
-  }
+		return false;
+	}
 
-  bool operator>(const Iterator &it) {
-    if (j_ > it.j_) {
-      return true;
-    } else if (j_ > it.j_) {
-      return false;
-    } else if (i_ > it.i_) // in this case, j_ == it.j_, hence comparison must
-                           // be done on i_
-    {
-      return true;
-    }
+	bool operator>=(const Iterator &it) {
+		if (!(*this < it)) {
+			return true;
+		}
 
-    return false;
-  }
+		return false;
+	}
 
-  bool operator>=(const Iterator &it) {
-    if (!(*this < it)) {
-      return true;
-    }
+	bool operator!=(const Iterator &it) {
+		if (i_ != it.i_ || j_ != it.j_)
+			return true;
+		else
+			return false;
+	}
 
-    return false;
-  }
+	Scalar &operator*() const { return (*p_mat_)(i_, j_); }
 
-  bool operator!=(const Iterator &it) {
-    if (i_ != it.i_ || j_ != it.j_)
-      return true;
-    else
-      return false;
-  }
+	Scalar *operator->() const { return &(*p_mat_)(i_, j_); }
 
-  Scalar &operator*() const { return (*p_mat_)(i_, j_); }
+	Iterator &operator++() {
+		if (i_ < rows_ - 1) // row increment
+		{
+			++i_;
+		} else // column increment
+		{
+			i_ = 0;
+			++j_;
+		}
+		return *this;
+	}
 
-  Scalar *operator->() const { return &(*p_mat_)(i_, j_); }
+	Iterator &operator++(int) {
+		if (i_ < rows_ - 1) // row increment
+		{
+			++i_;
+		} else // column increment
+		{
+			i_ = 0;
+			++j_;
+		}
+		return *this;
+	}
 
-  Iterator operator++() {
-    if (i_ < rows_ - 1) // row increment
-    {
-      ++i_;
-    } else // column increment
-    {
-      i_ = 0;
-      ++j_;
-    }
-    return Iterator(*this);
-  }
+	Iterator &operator--() {
+		if (i_ > 0) {
+			--i_;
+		} else {
+			i_ = rows_ - 1;
+			--j_;
+		}
+		return *this;
+	}
 
-  Iterator operator++(int) {
-    Iterator temp(*this);
+	int i_;
+	int j_;
+	int rows_;
+	Derived *p_mat_;
 
-    if (i_ < rows_ - 1) // row increment
-    {
-      ++i_;
-    } else // column increment
-    {
-      i_ = 0;
-      ++j_;
-    }
+	void posToIn(int pos, int &i, int &j) const {
+		std::div_t divresult;
+		divresult = std::div(pos, rows_);
 
-    return temp;
-  }
+		i = divresult.rem;
+		j = divresult.quot;
+	}
 
-  Iterator operator--() {
-    if (i_ > 0) {
-      --i_;
-    } else {
-      i_ = rows_ - 1;
-      --j_;
-    }
-    return Iterator(*this);
-  }
-
-  int i_;
-  int j_;
-  int rows_;
-  Derived *p_mat_;
-
-  void posToIn(int pos, int &i, int &j) const {
-    std::div_t divresult;
-    divresult = std::div(pos, rows_);
-
-    i = divresult.rem;
-    j = divresult.quot;
-  }
-
-  int pos() const { return j_ * rows_ + i_; }
+	int pos() const { return j_ * rows_ + i_; }
 };
 
 #endif // ITERATOR_H

--- a/MixtComp/src/lib/LinAlg/Iterator.h
+++ b/MixtComp/src/lib/LinAlg/Iterator.h
@@ -90,9 +90,7 @@ public:
 			return true;
 		} else if (j_ > it.j_) {
 			return false;
-		} else if (i_ < it.i_) // in this case, j_ == it.j_, hence comparison must
-							   // be done on i_
-		{
+		} else if (i_ < it.i_) { // in this case, j_ == it.j_, hence comparison must be done on i_
 			return true;
 		}
 
@@ -111,9 +109,7 @@ public:
 			return true;
 		} else if (j_ > it.j_) {
 			return false;
-		} else if (i_ > it.i_) // in this case, j_ == it.j_, hence comparison must
-							   // be done on i_
-		{
+		} else if (i_ > it.i_) { // in this case, j_ == it.j_, hence comparison must be done on i_
 			return true;
 		}
 

--- a/MixtComp/src/lib/LinAlg/Iterator.h
+++ b/MixtComp/src/lib/LinAlg/Iterator.h
@@ -36,9 +36,15 @@
  * mixt::Matrix.
  */
 
-class Iterator : public std::iterator<std::random_access_iterator_tag, Scalar,
-                                      int, Scalar *, Scalar &> {
+class Iterator {
 public:
+  // iterator traits
+  using iterator_category = std::random_access_iterator_tag;
+  using value_type = Scalar;
+  using difference_type = int;
+  using pointer = Scalar*;
+  using reference = Scalar&;
+
   Iterator(int i, int j, Derived &mat)
       : i_(i), j_(j), rows_(mat.rows()), p_mat_(&mat) {}
 

--- a/MixtComp/src/lib/LinAlg/Iterator.h
+++ b/MixtComp/src/lib/LinAlg/Iterator.h
@@ -53,6 +53,11 @@ public:
 		return *this;
 	}
 
+	Iterator& operator-=(int i) {
+		posToIn(pos() - i, i_, j_);
+		return *this;
+	}
+
 	Iterator operator-(int i) {
 		int posP, iP, jP;
 		posP = pos();

--- a/MixtComp/src/lib/LinAlg/Iterator.h
+++ b/MixtComp/src/lib/LinAlg/Iterator.h
@@ -138,6 +138,18 @@ public:
 		return *this;
 	}
 
+		Iterator& operator++(int) {
+		if (i_ < rows_ - 1) // row increment
+				{
+			++i_;
+		} else // column increment
+		{
+			i_ = 0;
+			++j_;
+		}
+		return *this;
+	}
+
 	Iterator& operator--() {
 		if (i_ > 0) {
 			--i_;

--- a/MixtComp/src/lib/LinAlg/Iterator.h
+++ b/MixtComp/src/lib/LinAlg/Iterator.h
@@ -145,7 +145,7 @@ public:
 
 	Scalar *operator->() const { return &(*p_mat_)(i_, j_); }
 
-	Iterator &operator++() {
+	Iterator operator++() {
 		if (i_ < rows_ - 1) // row increment
 		{
 			++i_;

--- a/MixtComp/src/lib/LinAlg/Iterator.h
+++ b/MixtComp/src/lib/LinAlg/Iterator.h
@@ -27,7 +27,7 @@
 // http://stackoverflow.com/questions/7758580/writing-your-own-stl-container/7759622#7759622
 
 //#include <iostream>
-//#include <iterator>
+#include <iterator>
 
 /**
  * Iterator class for Eigen::Matrix. This file is included from

--- a/MixtComp/src/lib/LinAlg/Iterator.h
+++ b/MixtComp/src/lib/LinAlg/Iterator.h
@@ -57,12 +57,6 @@ public:
 		  rows_(mat.rows()),
 		  p_mat_(&mat) {}
 
-	Iterator(const Iterator &other)
-		: i_(other.i_),
-		  j_(other.j_),
-		  rows_(other.rows_),
-		  p_mat_(other.p_mat_) {}
-
 	Iterator operator+(int i) {
 		int posP, iP, jP;
 		posP = pos();

--- a/MixtComp/src/lib/LinAlg/Iterator.h
+++ b/MixtComp/src/lib/LinAlg/Iterator.h
@@ -23,159 +23,161 @@
 #ifndef ITERATOR_H
 #define ITERATOR_H
 
-// reference on what needs to be implemented for STL iterators: http://stackoverflow.com/questions/7758580/writing-your-own-stl-container/7759622#7759622
+// reference on what needs to be implemented for STL iterators:
+// http://stackoverflow.com/questions/7758580/writing-your-own-stl-container/7759622#7759622
 
-//#include <iostream>
-//#include <iterator>
+// #include <iostream>
+// #include <iterator>
 
 /**
- * Iterator class for Eigen::Matrix. This file is included from EigenMatrixBaseAddons.h and the Iterator class is therefore not a member of the namespace mixt::.
- * It is accessible in MixtComp through the derived class mixt::Matrix.
+ * Iterator class for Eigen::Matrix. This file is included from
+ * EigenMatrixBaseAddons.h and the Iterator class is therefore not a member of
+ * the namespace mixt::. It is accessible in MixtComp through the derived class
+ * mixt::Matrix.
  */
 
-class Iterator: public std::iterator<std::random_access_iterator_tag, Scalar,
-		int, Scalar*, Scalar&> {
+class Iterator : public std::iterator<std::random_access_iterator_tag, Scalar,
+                                      int, Scalar *, Scalar &> {
 public:
-	Iterator(int i, int j, Derived& mat) :
-			i_(i), j_(j), rows_(mat.rows()), p_mat_(&mat) {
-	}
+  Iterator(int i, int j, Derived &mat)
+      : i_(i), j_(j), rows_(mat.rows()), p_mat_(&mat) {}
 
-	Iterator operator+(int i) {
-		int posP, iP, jP;
-		posP = pos();
-		posP += i;
-		posToIn(posP, iP, jP);
-		return Iterator(iP, jP, *p_mat_);
-	}
+  Iterator(const Iterator &other)
+      : i_(other.i_), j_(other.j_), rows_(other.rows_), p_mat_(other.p_mat_) {}
 
-	Iterator& operator+=(int i) {
-		posToIn(pos() + i, i_, j_);
-		return *this;
-	}
+  Iterator operator+(int i) {
+    int posP, iP, jP;
+    posP = pos();
+    posP += i;
+    posToIn(posP, iP, jP);
+    return Iterator(iP, jP, *p_mat_);
+  }
 
-	Iterator& operator-=(int i) {
-		posToIn(pos() - i, i_, j_);
-		return *this;
-	}
+  Iterator &operator+=(int i) {
+    posToIn(pos() + i, i_, j_);
+    return *this;
+  }
 
-	Iterator operator-(int i) {
-		int posP, iP, jP;
-		posP = pos();
-		posP -= i;
-		posToIn(posP, iP, jP);
-		return Iterator(iP, jP, *p_mat_);
-	}
+  Iterator &operator-=(int i) {
+    posToIn(pos() - i, i_, j_);
+    return *this;
+  }
 
-	int operator-(const Iterator& it) {
-		return pos() - it.pos();
-	}
+  Iterator operator-(int i) {
+    int posP, iP, jP;
+    posP = pos();
+    posP -= i;
+    posToIn(posP, iP, jP);
+    return Iterator(iP, jP, *p_mat_);
+  }
 
-	bool operator<(const Iterator& it) {
-		if (j_ < it.j_) {
-			return true;
-		} else if (j_ > it.j_) {
-			return false;
-		} else if (i_ < it.i_) // in this case, j_ == it.j_, hence comparison must be done on i_
-				{
-			return true;
-		}
+  int operator-(const Iterator &it) { return pos() - it.pos(); }
 
-		return false;
-	}
+  bool operator<(const Iterator &it) {
+    if (j_ < it.j_) {
+      return true;
+    } else if (j_ > it.j_) {
+      return false;
+    } else if (i_ < it.i_) // in this case, j_ == it.j_, hence comparison must
+                           // be done on i_
+    {
+      return true;
+    }
 
-	bool operator==(const Iterator& it) {
-		if (i_ == it.i_ && j_ == it.j_)
-			return true;
-		else
-			return false;
-	}
+    return false;
+  }
 
-	bool operator>(const Iterator& it) {
-		if (j_ > it.j_) {
-			return true;
-		} else if (j_ > it.j_) {
-			return false;
-		} else if (i_ > it.i_) // in this case, j_ == it.j_, hence comparison must be done on i_
-				{
-			return true;
-		}
+  bool operator==(const Iterator &it) {
+    if (i_ == it.i_ && j_ == it.j_)
+      return true;
+    else
+      return false;
+  }
 
-		return false;
-	}
+  bool operator>(const Iterator &it) {
+    if (j_ > it.j_) {
+      return true;
+    } else if (j_ > it.j_) {
+      return false;
+    } else if (i_ > it.i_) // in this case, j_ == it.j_, hence comparison must
+                           // be done on i_
+    {
+      return true;
+    }
 
-	bool operator>=(const Iterator& it) {
-		if (!(*this < it)) {
-			return true;
-		}
+    return false;
+  }
 
-		return false;
-	}
+  bool operator>=(const Iterator &it) {
+    if (!(*this < it)) {
+      return true;
+    }
 
-	bool operator!=(const Iterator& it) {
-		if (i_ != it.i_ || j_ != it.j_)
-			return true;
-		else
-			return false;
-	}
+    return false;
+  }
 
-	Scalar& operator*() const {
-		return (*p_mat_)(i_, j_);
-	}
+  bool operator!=(const Iterator &it) {
+    if (i_ != it.i_ || j_ != it.j_)
+      return true;
+    else
+      return false;
+  }
 
-	Scalar* operator->() const {
-		return &(*p_mat_)(i_, j_);
-	}
+  Scalar &operator*() const { return (*p_mat_)(i_, j_); }
 
-	Iterator& operator++() {
-		if (i_ < rows_ - 1) // row increment
-				{
-			++i_;
-		} else // column increment
-		{
-			i_ = 0;
-			++j_;
-		}
-		return *this;
-	}
+  Scalar *operator->() const { return &(*p_mat_)(i_, j_); }
 
-		Iterator& operator++(int) {
-		if (i_ < rows_ - 1) // row increment
-				{
-			++i_;
-		} else // column increment
-		{
-			i_ = 0;
-			++j_;
-		}
-		return *this;
-	}
+  Iterator operator++() {
+    if (i_ < rows_ - 1) // row increment
+    {
+      ++i_;
+    } else // column increment
+    {
+      i_ = 0;
+      ++j_;
+    }
+    return Iterator(*this);
+  }
 
-	Iterator& operator--() {
-		if (i_ > 0) {
-			--i_;
-		} else {
-			i_ = rows_ - 1;
-			--j_;
-		}
-		return *this;
-	}
+  Iterator operator++(int) {
+    Iterator temp(*this);
 
-	int i_;
-	int j_;
-	int rows_;
-	Derived* p_mat_;
+    if (i_ < rows_ - 1) // row increment
+    {
+      ++i_;
+    } else // column increment
+    {
+      i_ = 0;
+      ++j_;
+    }
 
-	void posToIn(int pos, int& i, int& j) const {
-		std::div_t divresult;
-		divresult = std::div(pos, rows_);
+    return temp;
+  }
 
-		i = divresult.rem;
-		j = divresult.quot;
-	}
+  Iterator operator--() {
+    if (i_ > 0) {
+      --i_;
+    } else {
+      i_ = rows_ - 1;
+      --j_;
+    }
+    return Iterator(*this);
+  }
 
-	int pos() const {
-		return j_ * rows_ + i_;
-	}
+  int i_;
+  int j_;
+  int rows_;
+  Derived *p_mat_;
+
+  void posToIn(int pos, int &i, int &j) const {
+    std::div_t divresult;
+    divresult = std::div(pos, rows_);
+
+    i = divresult.rem;
+    j = divresult.quot;
+  }
+
+  int pos() const { return j_ * rows_ + i_; }
 };
 
 #endif // ITERATOR_H

--- a/MixtComp/src/lib/LinAlg/Iterator.h
+++ b/MixtComp/src/lib/LinAlg/Iterator.h
@@ -46,11 +46,22 @@ public:
 	using pointer = Scalar *;
 	using reference = Scalar &;
 
+	int i_;
+	int j_;
+	int rows_;
+	Derived *p_mat_;
+
 	Iterator(int i, int j, Derived &mat)
 		: i_(i),
 		  j_(j),
 		  rows_(mat.rows()),
 		  p_mat_(&mat) {}
+
+	Iterator(const Iterator &other)
+		: i_(other.i_),
+		  j_(other.j_),
+		  rows_(other.rows_),
+		  p_mat_(other.p_mat_) {}
 
 	Iterator operator+(int i) {
 		int posP, iP, jP;
@@ -60,12 +71,12 @@ public:
 		return Iterator(iP, jP, *p_mat_);
 	}
 
-	Iterator &operator+=(int i) {
+	Iterator operator+=(int i) {
 		posToIn(pos() + i, i_, j_);
 		return *this;
 	}
 
-	Iterator &operator-=(int i) {
+	Iterator operator-=(int i) {
 		posToIn(pos() - i, i_, j_);
 		return *this;
 	}
@@ -146,7 +157,8 @@ public:
 		return *this;
 	}
 
-	Iterator &operator++(int) {
+	Iterator operator++(int) {
+		Iterator temp(*this);
 		if (i_ < rows_ - 1) // row increment
 		{
 			++i_;
@@ -155,10 +167,10 @@ public:
 			i_ = 0;
 			++j_;
 		}
-		return *this;
+		return temp;
 	}
 
-	Iterator &operator--() {
+	Iterator operator--() {
 		if (i_ > 0) {
 			--i_;
 		} else {
@@ -167,11 +179,6 @@ public:
 		}
 		return *this;
 	}
-
-	int i_;
-	int j_;
-	int rows_;
-	Derived *p_mat_;
 
 	void posToIn(int pos, int &i, int &j) const {
 		std::div_t divresult;

--- a/MixtComp/src/lib/LinAlg/Iterator.h
+++ b/MixtComp/src/lib/LinAlg/Iterator.h
@@ -36,10 +36,16 @@
  * mixt::Matrix.
  */
 
-class Iterator : public std::iterator<
-					 std::random_access_iterator_tag, Scalar,
-					 int, Scalar *, Scalar &> {
+class Iterator {
+
 public:
+	// iterator traits
+	using iterator_category = std::random_access_iterator_tag;
+	using value_type = Scalar;
+	using difference_type = int;
+	using pointer = Scalar *;
+	using reference = Scalar &;
+
 	Iterator(int i, int j, Derived &mat)
 		: i_(i),
 		  j_(j),

--- a/MixtComp/src/lib/LinAlg/Iterator.h
+++ b/MixtComp/src/lib/LinAlg/Iterator.h
@@ -27,7 +27,7 @@
 // http://stackoverflow.com/questions/7758580/writing-your-own-stl-container/7759622#7759622
 
 //#include <iostream>
-#include <iterator>
+// #include <iterator>
 
 /**
  * Iterator class for Eigen::Matrix. This file is included from

--- a/MixtComp/src/lib/Statistic/MultinomialStatistic.cpp
+++ b/MixtComp/src/lib/Statistic/MultinomialStatistic.cpp
@@ -25,16 +25,9 @@
 
 namespace mixt {
 
-MultinomialStatistic::MultinomialStatistic() :
-		rng_(seed(this)), uni_(0., 1.), g_(rng_), generator_(rng_, uni_) {
-}
-;
-
 int MultinomialStatistic::sampleInt(int low, int high) {
-	boost::random::uniform_int_distribution<> uni(low, high);
-	boost::variate_generator<boost::random::mt19937&,
-			boost::random::uniform_int_distribution<> > generator(rng_, uni);
-	int x = generator();
+	std::uniform_int_distribution<int> uni_(low, high);
+	int x = uni_(rng_);
 	return x;
 }
 

--- a/MixtComp/src/lib/Statistic/MultinomialStatistic.h
+++ b/MixtComp/src/lib/Statistic/MultinomialStatistic.h
@@ -25,22 +25,35 @@
 
 #include <iostream>
 #include <ctime>
-#include <boost/random.hpp>
-#include <boost/range/algorithm/random_shuffle.hpp>
+#include <random>
 #include <Various/Constants.h>
 #include "../LinAlg/LinAlg.h"
 #include "../IO/IO.h"
+#include "Statistic/RNG.h"
 
 namespace mixt {
 
 class MultinomialStatistic {
 
+private:
+/** Random number generator */
+std::mt19937 rng_;
+
+std::uniform_real_distribution<Real> uni_;
+
+// std::random_number_generator<std::mt19937> g_;
+
+// boost::variate_generator<boost::random::mt19937&,
+// 		boost::random::uniform_real_distribution<> > generator_;
+
 public:
-	MultinomialStatistic();
+	MultinomialStatistic()
+	: rng_(seed(this)),
+	uni_(0.0, 1.0) {}
 
 	/** Sample a value from a binomial law with  */
 	int sampleBinomial(Real proportion) {
-		if (generator_() < proportion) {
+		if (uni_(rng_) < proportion) {
 			return 1;
 		} else {
 			return 0;
@@ -50,7 +63,7 @@ public:
 	/** Sample a value from a multinomial law with coefficient of modalities provided */
 	template<typename T>
 	int sample(const T& proportion) {
-		Real x = generator_();
+		Real x = uni_(rng_);
 
 		Real cumProb = 0.; // cumulative probability
 		int index = 0;
@@ -67,7 +80,7 @@ public:
 
 	template<typename T>
 	void shuffle(T& data) {
-		boost::range::detail::random_shuffle(data.begin(), data.end(), g_);
+		std::shuffle(data.begin(), data.end(), rng_);
 	}
 
 	template<typename T>
@@ -87,18 +100,6 @@ public:
 	 * @return integer uniformly sampled from [low:high]
 	 */
 	int sampleInt(int low, int high);
-
-private:
-	/** Random number generator */
-	boost::random::mt19937 rng_;
-
-	boost::random::uniform_real_distribution<> uni_;
-
-	boost::random_number_generator<boost::mt19937> g_;
-
-	boost::variate_generator<boost::random::mt19937&,
-			boost::random::uniform_real_distribution<> > generator_;
-
 };
 
 } // namespace mixt

--- a/MixtComp/src/lib/Statistic/MultinomialStatistic.h
+++ b/MixtComp/src/lib/Statistic/MultinomialStatistic.h
@@ -49,7 +49,7 @@ std::uniform_real_distribution<Real> uni_;
 public:
 	MultinomialStatistic()
 	: rng_(seed(this)),
-	uni_(0.0, 1.0) {}
+	  uni_(0.0, 1.0) {}
 
 	/** Sample a value from a binomial law with  */
 	int sampleBinomial(Real proportion) {

--- a/MixtComp/src/utest/Rank/UTestRankISRIndividual.cpp
+++ b/MixtComp/src/utest/Rank/UTestRankISRIndividual.cpp
@@ -69,7 +69,7 @@ TEST(RankISRIndividual, switchRepresentation) {
 TEST(RankISRIndividual, xGenP05) {
 	int nbPos = 4;
 	int nbSample = 15000;
-	Real tolerance = 1.e-3;
+	Real tolerance = 1.5e-3;
 	int nbE = fac(nbPos);
 	Real logProba = -std::log(nbE);
 

--- a/RMixtCompIO/DESCRIPTION
+++ b/RMixtCompIO/DESCRIPTION
@@ -1,8 +1,8 @@
 Package: RMixtCompIO
 Type: Package
 Title: Minimal Interface of the C++ 'MixtComp' Library for Mixture Models with Heterogeneous and (Partially) Missing Data
-Version: 4.0.8
-Date: 2022-10-18
+Version: 4.0.9
+Date: 2023-02-11
 Authors@R: c(person("Vincent", "Kubicki", role = "aut"),
              person("Christophe", "Biernacki", role = "aut"),
              person("Quentin", "Grimonprez", role = c("aut", "cre"), email = "quentingrim@yahoo.fr"),
@@ -23,5 +23,5 @@ BugReports: https://github.com/modal-inria/MixtComp/issues
 Imports: Rcpp, doParallel, foreach
 Suggests: Rmixmod, testthat, RInside, xml2
 LinkingTo: Rcpp, RcppEigen, BH
-RoxygenNote: 7.2.1
+RoxygenNote: 7.2.3
 Encoding: UTF-8

--- a/RMixtCompIO/updateLib.py
+++ b/RMixtCompIO/updateLib.py
@@ -1,7 +1,7 @@
 #! /usr/local/bin/python3
 
 # MixtComp version 4 - july 2019
-# Copyright (C) Inria - Université de Lille - CNRS 
+# Copyright (C) Inria - Université de Lille - CNRS
 
 import glob
 import subprocess
@@ -32,7 +32,7 @@ makevarsStr = r'''# MixtComp version 4  - july 2019
 %s
 
 PKG_CXXFLAGS = -Ilib -Ilib/LinAlg -DEIGEN_MATRIXBASE_PLUGIN=\"EigenMatrixBaseAddons.h\" -Ioptim/include
-CXX_STD = CXX11
+CXX_STD = CXX17
 
 PKG_LIBS = $(LIB_OFILES)
 

